ruby_programming/git/advanced_git_lesson.md

```diff
@@ -0,0 +1,182 @@
+
+Now that you've had a second to gather your thoughts and attempt to wrap your head around this concept, it might help to go back and look at a concrete example of pointers we used in this lesson. Let's think back to our use of `git rebase -i HEAD~3`. If you can remember, this command let us edit the last 3 commits. Do you have any guesses on how git knew which 3 commits to edit? That's right, by using pointers! We start at HEAD, a special pointer for keeping track of the branch you're currently on. HEAD points to our most recent commit in the current branch. That commit pointed to the commit made directly before it, which we can call commit 2. That commit does the exact same, and points to the commit before it, which we can call commit 3. That's how `git rebase -i HEAD~3` starts with a HEAD pointer, and then uses more pointers to see which three commits to edit. 
+
+You might be feeling overwhelmed at this point, so let's recap what we've learned. A branch is simply a pointer to a single commit. A commit is a snapshot, and a pointer at the commit directly behind it in history. That's it!
+
+### Additional Resources
+This section contains helpful links to other content. It isn't required, so consider it supplemental.
+* Read [Git Cheat Sheet](https://www.atlassian.com/git/tutorials/atlassian-git-cheatsheet) if you need a reference sheet.
+* Watch [Using Rebase & Merge](https://www.youtube.com/watch?v=f1wnYdLEpgI) for an example on how to use both rebase and merge.
+* Read [Branches in a Nutshell](https://git-scm.com/book/en/v2/Git-Branching-Branches-in-a-Nutshell) if you want an even deeper dive into Branches.
+* Read [this chapter on Rebasing](https://git-scm.com/book/en/v2/Git-Branching-Rebasing) for an even deeper dive into Rebasing.
+
+### Knowledge Check
+This section contains questions for you to check your understanding of this lesson. If you're having trouble answering the questions below on your own, review the material above to find the answer.
+
+* <a class='knowledge-check-link' href='https://git-scm.com/book/en/v2/Git-Basics-Undoing-Things'>How can you amend your last commit?</a>
+* <a class='knowledge-check-link' href='https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History'>What are some different ways to rewrite history?</a>
+* <a class='knowledge-check-link' href='#force-with-lease-knowledge-check'>What is a safe way to push history changes to a remote repository?</a>
+* <a class='knowledge-check-link' href='#dangers-knowledge-check'>What are the dangers of history-changing operations</a>
+* <a class='knowledge-check-link' href='#best-practices-knowledge-check'>What are best practices of history-changing operations</a>
```

Comment:
I don't think that the Knowledge Check ids need to include "knowledge check" in them.